### ISSUE TITLE
This Plugin is Built Around Ember Simple Auth - via the OAuth 2 Authenticator, Shouldn't We Make the TokenLoginLogoutAuthenticate More Closely Match the OAuth 2 Spec?

### ISSUE THREAD

**Issue by jtyost2** [loadsys/CakePHP-Stateless-Auth#12]
I recognize this is a large task and maybe this is the argument for not publishing the plugin, given we are basically implementing part of a standard without all of it in place.

I'm not even really arguing for this to support all features of OAuth 2 or even consider it 100% compatible, but the use for this in FM is a OAuth 2 style flow - so shouldn't we at least get closer to the OAuth 2 spec?

Keeping in mind there are a lot of packages out there now that do OAuth servers in PHP: http://oauth.net/2/#implementations

Reference to Ember Simple Auth: https://github.com/simplabs/ember-simple-auth
Ref to Ember Simple Auth OAuth 2: https://github.com/simplabs/ember-simple-auth/tree/master/packages/ember-simple-auth-oauth2
Ref to OAuth 2 Spec: http://oauth.net/2/

**Comment by jtyost2** [comment]
It seems like TokenAuthenticate doesn't really have a flow for matching the spec, given that there is no login/logout flow, so that's exempt from my point about OAuth 2.

**Comment by beporter** [comment]
Geez. After [reading up a bit about OAuth2](http://en.wikipedia.org/wiki/OAuth), why again are we using it? Why is anyone?

**Comment by jtyost2** [comment]
I'm not sure you're joking about this or not.

Because:

* OAuth 1 is so much worse

The only other standards for Authorization flows are so insecure it's laughable.

* [https://en.wikipedia.org/wiki/Basic_access_authentication](https://en.wikipedia.org/wiki/Basic_access_authentication)
* [https://en.wikipedia.org/wiki/Digest_access_authentication](https://en.wikipedia.org/wiki/Digest_access_authentication)

OAuth has some issues and is overly complex in some places, I don't disagree. It's also a standards that is designed to capture as many use cases as possible to be "the authorization" when you aren't directly logging into a app directly.

It's also why while: Facebook, Twitter, Google, etc all use it they typically don't follow the spec 100% - they all do slight variations for their implementation for issues they see with it.

In short, are there issues with it yes. Is there a better spec for being able to implement Authorization flows on the web, not really.

**Comment by jtyost2** [comment]
As discussed, suggested areas of improvement for meeting the OAuth 2 spec:

* Validating the `grant_type`
* Validate the `client_id`
* Validate the `client_secret`
* Ensure SSL on requests
* Provide valid response including things like the `expires` etc
